pkg/oci/authn/gke.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 The Falco Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authn
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"oras.land/oras-go/v2/registry/remote/auth"
+
+	"github.com/falcosecurity/falcoctl/internal/config"
+)
+
+// GkeClientCredentialsStore provides credential retrieval for gke client credentials.
+type GkeClientCredentialsStore struct {
+	Gke2TokenSources map[string]oauth2.TokenSource
+}
+
+// NewOauthClientCredentialsStore creates a new Gke client credential store.
+func NewGkeClientCredentialsStore() GkeClientCredentialsStore {
+	return GkeClientCredentialsStore{
+		Gke2TokenSources: make(map[string]oauth2.TokenSource),
+	}
+}
+
+// Credential retrieves a valid access token auth credential for the given registry.
+func (o *GkeClientCredentialsStore) Credential(ctx context.Context, reg string) (auth.Credential, error) {
+	tokenSource, exists := o.Gke2TokenSources[reg]
+	// if we did not already load a token source for this registry check the config file
+	if !exists {
+		gkeAuths, err := config.Gkes()
+		if err != nil {
+			return auth.EmptyCredential, fmt.Errorf("unable to retrieve gke credentials %w", err)
+		}
+
+		for _, gke := range gkeAuths {
+			if gke.Registry == reg {
+				tokenSource, err = google.DefaultTokenSource(ctx)
```

Comment:
Nice work with the PR!

Might be premature since its still WIP, but when looking over it now I was wondering if we actually need one token source per registry here. Since unless I am wrong with workload identity we get only one identity per exec env anyway. Unless we add support for specifying different credential files per registry. So simply caching that workload identity is configured for the registry and then re-using the same token source for all should work. 🤔 💭

Reply:
Hi max! Thank you very much for your review!
Indeed, I definitely knew some things are still to be disccused/not in their best form for now, and you pointed the finger at one (or two) of them :sweat_smile: 

First, I agree with you, no need to cache the Workload identity source, since it won't change for the same workload :sweat_smile: .
But here, I still used the `google.DefaultTokenSource(ctx)` instead of the `google.ComputeTokenSource("")` which only uses the GCE metadata server (or Workload Identity when that's the case). And this `DefaultTokenSource` (see [here](https://github.com/golang/oauth2/blob/master/google/default.go#L109) can use other sources when available (`GOOGLE_APPLICATION_CREDENTIALS` for instance).
Hence why I first saved the found tokenSource. But even right now I still doubt it's a good idea^^

Reply:
Ah no what I mean here is that we only need one DefaulTokenSource for all the registries. So we only need to load it once and then can re-use it for any registry that's set to workload identity/application default credentials